### PR TITLE
FEATURE: prettify slug generation and specs

### PR DESCRIPTION
**Pull request by fantasticfears**
Regards the null out after changing the site settings. I am afraid it's expensive. 2 options, 1st trigger a regular job. 2nd regular cleaner like reindexer.

### CODE REVIEW

**Comment by fantasticfears** [comment]
@SamSaffron I have a disparity here. https://travis-ci.org/discourse/discourse/builds/267106586
In my local chrome, it works with this changeset even it breaks CI. Chrome always send the percent encoded string. Rails encode the parameter wrong. So our persistent strategy in this changeset which stores the percent encoded slug does not work.
So hack slug parameters?

**Comment by SamSaffron** [comment]
@tgxworld will review this, but we want the rails 5 upgrade done first

**Comment by fantasticfears** [comment]
As an aside, Rails 5.1 returns the UTF-8 string. So I'll revisit this after our Rails 5 upgrade.

**Comment by tgxworld** [comment]
Hi @fantasticfears, Rails 5.1 is more or less stable. Let me know if this PR needs to be updated.

**Comment by fantasticfears** [comment]
Sure, I am going to review this PR.

**Comment by SamSaffron** [comment]
not following, do we even need any changes here post 5.1 ... I thought the framework fixed it.

**Comment by fantasticfears** [comment]
If you no longer need to persist it to encoded, we should put on hold on this... But I will extract the test though.

File: lib/slug.rb [hunk 1]
```diff
@@ -13,11 +14,12 @@ def self.for(string, default = 'topic')
       end
     # Reject slugs that only contain numbers, because they would be indistinguishable from id's.
     slug = (slug =~ /[^\d]/ ? slug : '')
+    slug = slug.length >= MAX_SLUG_LENGTH ? '' : slug
```
Comment (SamSaffron) [inline lib/slug.rb]: are we sure about this, why not truncate?

Comment (fantasticfears) [reply lib/slug.rb]: I was definitely out of my mind...will fix

**Comment by fantasticfears** [comment]
rebased and spec fixed

**Comment by SamSaffron** [comment]
I am just so confused here, what specifically is this PR fixing now?

**Comment by fantasticfears** [comment]
Sorry, I should change the title and leave comments as well.

Now the encoded generator will try to downcase and restrict slug string which is longer than a very large number 255.
Moreover, previous encoded generator may leave out a trailing dash. This is fixed in this PR too.

**Comment by tgxworld** [comment]
Merged in https://github.com/discourse/discourse/commit/7c3123a2dd814663c86cd6b4ed3619ca3c02aab4

Sorry for the delay @fantasticfears

**Pull request closed without merging**